src/test/ui/consts/const_in_pattern/reject_non_partial_eq.rs

```diff
@@ -0,0 +1,32 @@
+// This test is illustrating the difference between how failing to derive
+// `PartialEq` is handled compared to failing to implement it at all.
+
+// See also RFC 1445
+
+#[derive(PartialEq, Eq)]
+struct Structural(u32);
+
+struct NoPartialEq(u32);
+
+struct NoDerive(u32);
+
+// This impl makes NoDerive irreflexive.
+impl PartialEq for NoDerive { fn eq(&self, _: &Self) -> bool { false } }
+
+impl Eq for NoDerive { }
+
+const NO_DERIVE_NONE: Option<NoDerive> = None;
+const NO_PARTIAL_EQ_NONE: Option<NoPartialEq> = None;
+
+fn main() {
+    match None {
+        NO_DERIVE_NONE => println!("NO_DERIVE_NONE"),
+        _ => panic!("whoops"),
+    }
+
+    match None {
+        NO_PARTIAL_EQ_NONE => println!("NO_PARTIAL_EQ_NONE"),
+        //~^ ERROR must be annotated with `#[derive(PartialEq, Eq)]`
```

Comment:
This is fine for now, since, if I recall correctly, we justify this based on it never having compiled in the past (right?)

But if we are committing to the Qualif based machinery for the structural-match checking, then maybe it would make sense to treat the `NO_PARTIAL_EQ_NONE` case the same as the `NO_DERIVE_NONE` case, no?

(This is more a question for @rust-lang/lang , rather than @ecstatic-morse , though maybe @ecstatic-morse has a killer example of why the semantics here for `NO_PARTIAL_EQ_NONE` is and always will be the right thing.)

Reply:
My understanding is that, if we allow this to compile, we lose the ability to use semantic equality instead of structural equality for pattern matching. It's possible that this ship has sailed, but we should have that discussion. If we are willing to punt on the notion of semantic equality for pattern matching, I think this would indeed be a valid choice.

Reply:
Oh right, of course; I just got finished writing a comment elsewhere noting that we had not yet commited to semantic equality for the semantics here, *of course* we cannot accept the code that leaves out the `PartialEq` implementation.